src/wasm/wasm-type.cpp

```diff
@@ -32,25 +33,44 @@ struct TypeInfo {
     TupleKind,
     RefKind,
     RttKind,
+    TempRefKind,
+    TempRttKind,
   } kind;
   struct Ref {
     HeapType heapType;
     bool nullable;
   };
+  struct TempRef {
+    HeapType* heapType;
+    bool nullable;
+  };
+  struct TempRtt {
+    uint32_t depth;
+    HeapType* heapType;
+  };
```

Comment:
This makes me wonder once again if maybe HeapTypes should always be by a pointer. That is, that on line 37/40 we'd have `HeapType*` instead of `HeapType`.

I may be missing something, but it seems like that would make it simpler to implement the TypeBuilder, as we wouldn't need to add `TempRef` and `TempRtt`. And, separately, it would also avoid the risk of copies being passed around.

If memory safety is a concern here, then our usage pattern is very simple - allocate types when creating a wasm, and then use them. We could just store them in a global arena that is freed on process exit, like strings are. Note that that is a separate question from whether we intern them or not (which I am not sure about either way).

One of my concerns about `TempRef` and `TempRtt`, more specifically, is that they end up intruding in various places throughout this file. If somehow the TypeBuilder could encapsulate that complexity into itself then that would be another option to think about.

Reply:
I'll experiment with this.